dnscrypt.c

```diff
@@ -3,19 +3,29 @@
 const KeyPair *
 find_keypair(const struct context *c,
              const unsigned char magic_query[DNSCRYPT_MAGIC_HEADER_LEN],
-             const size_t dns_query_len)
+             const size_t dns_query_len,
+             bool *use_xchacha20)
 {
+    unsigned char magic_query_ref[DNSCRYPT_MAGIC_HEADER_LEN];
     const KeyPair *keypairs = c->keypairs;
     size_t i;
 
     if (dns_query_len <= DNSCRYPT_QUERY_HEADER_SIZE) {
         return NULL;
     }
+    *use_xchacha20 = 0;
     for (i = 0U; i < c->keypairs_count; i++) {
-        if (memcmp(keypairs[i].crypt_publickey, magic_query,
-                   DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+        memcpy(magic_query_ref, keypairs[i].crypt_publickey, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
             return &keypairs[i];
         }
+#ifdef HAVE_CRYPTO_BOX_CURVE25519XCHACHA20POLY1305_OPEN_EASY
+        sodium_increment(magic_query_ref, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+            *use_xchacha20 = 1;
```

Comment:
I was not too concerned about collision, as you said, they are very unlikely. Also as you said, the client_magic is totally arbitrary as long as no 2 certs have the same one.
Anyhow, in the end, what I was looking a bit more into was to find a way to infer the `es_version` from the actual certificate rather than arbitrarily from the fact that `magic_query == pk[0:8] + 1`.

I tried https://gist.github.com/chantra/9e175e78e5cb2eb9a570e356f1298b11, which works but I failed at taking into account that the same key may be used to generate 1 cert for xsalsa20 and 1 cert for xchacha20, in which case solely matching the public key is not going to cut it.